libs2eplugins/src/s2e/Plugins/Models/BaseFunctionModels.cpp

```diff
@@ -692,42 +692,67 @@ bool BaseFunctionModels::strcatHelper(S2EExecutionState *state, const uint64_t s
+    if (!findNullCharWithWidth(state, haystackAddr, haystackLen, byte_width) ||
+        !findNullCharWithWidth(state, needleAddr, needleLen, byte_width)) {
         getWarningsStream(state) << "Failed to find nullptr in haystack or needle\n";
         return false;
     }
 
-    const Expr::Width ret_width = state->getPointerSize() * CHAR_BIT;
     if (needleLen == 0) {
-        retExpr = E_CONST(haystackAddr, ret_width);
-        return true;
+        retExpr = E_CONST(haystackAddr, state->getPointerSize() * CHAR_BIT);
+        return true;  // Immediate return if needle is an empty string
     }
 
-    const ref<Expr> nullExpr = E_CONST(0, ret_width);
+    const Expr::Width pointerWidth = state->getPointerSize() * CHAR_BIT;
+    const ref<Expr> nullExpr = E_CONST(0, pointerWidth);
 
-    ref<Expr> finalExpr = nullExpr; // Start with the assumption that we will not find the needle
-    for (size_t i = 0; i <= haystackLen - needleLen; i += byte_width) {
-        uint64_t strAddrs[2] = {haystackAddr + i, needleAddr};
-        ref<Expr> memcmpResult;
-
-        // Use strncmpHelper to compare the substring of haystack with needle
-        if (!memcmpHelper(state, strAddrs, needleLen, memcmpResult)) {
-            getWarningsStream(state) << "Failed to compare substring of haystack with needle " << needleLen << "\n";
+    std::vector<size_t> badCharSkip(256, needleLen);  // Bad character skip array
+    uint8_t charByte = 0;  // Temporary variable for reading bytes
+    for (size_t i = 0; i < needleLen - 1; ++i) {
+        uint64_t readAddr = needleAddr + i * byte_width;
+        if (!state->mem()->read(readAddr, &charByte, VirtualAddress, true)) {
+            getWarningsStream(state) << "Failed to read byte at address " << hexval(readAddr) << "\n";
             return false;
         }
+        badCharSkip[charByte] = needleLen - 1 - i;
+    }
+
+    size_t i = 0;  // Position in haystack
+    auto solver = state->solver();
+    while (i <= haystackLen - needleLen) {
+        uint64_t currentHaystackAddr = haystackAddr + i * byte_width;
+        uint64_t strAddrs[2] = {currentHaystackAddr, needleAddr};
+        ref<Expr> cmpResult;
+
+        if (strncmpHelper(state, strAddrs, needleLen, cmpResult)) {
+            // incrementally adding the state constriant based on solver feedback
+            ref<Expr> findNeedle = E_EQ(cmpResult, E_CONST(0, Expr::Int32));
+            Query query(state->constraints(), findNeedle);
+            bool truth;
+            if (solver->mayBeTrue(query,truth)) {
+                retExpr = E_CONST(haystackAddr + i, state->getPointerSize() * CHAR_BIT);
+                return true;
+            }
+        }
 
-        // Build the conditional expression for each possible starting position of needle in haystack
-        finalExpr = E_ITE(E_EQ(memcmpResult, E_CONST(0, Expr::Int32)), E_CONST(haystackAddr + i, ret_width), finalExpr);
+        uint64_t nextCharAddr = currentHaystackAddr + needleLen * byte_width;
```

Comment:
Is this the standard way of implementing the Bad Character Heuristic? What if the nextChar does not exist in the needle? It seems that it will result in an infinite loop.

Consider the following:

```
abcdefghijkl -> haystack
cbc -> needle
```

In the first iteration of the while loop, strncmpHelper will return false, and `nextCharAddr` will be set to haystackAddr + 3, which points to `d`. Since 'd' does not exist in the needle, badCharSkip['d'] is not set and thus equals 0. As a result, at line 747, `i += 0`, and the function will enter an infinite loop.

I prefer to copy a well-tested implementation (e.g., KMP) from online repos and use that instead of recreating our wheels, which could be error-prone.

@Changochen What do you think?

Reply:
What is the reference for the standard? This is my first time hearing bad char heuristic.

Reading this implementation, I think the important trick is that we stop if once we find there is a potential match. That should be easily done by keeping only this part:

```
 ref<Expr> findNeedle = E_EQ(cmpResult, E_CONST(0, Expr::Int32));
            Query query(state->constraints(), findNeedle);
            bool truth;
            if (solver->mayBeTrue(query,truth)) {
                retExpr = E_CONST(haystackAddr + i, state->getPointerSize() * CHAR_BIT);
                return true;
            }
```

KMP is not heuristic but has proven bound. What complexity reduction we have by using the bad char hearustic?

Reply:
@Changochen FYI the badChar Heuristic is a well known string search algorithm called Boyer-Moore string search algorithm. https://en.wikipedia.org/wiki/Boyer%E2%80%93Moore_string-search_algorithm

Reply:
> Yes, I follow the standard Bad Char Heuristic. Actually, all of the non-existing char for needle the badCharSkip is initialized to needleLen on line 713. If 'd' does not exists in badCharSkip, needleLen will be returned and for size of needleLen chars will be skipped.

Oh sorry I was blind. In this case, my example won't cause an infinite loop.

Reply:
The other trick is that if underlying haystack has concrete values/or we have concretized to some values, we can speed up the searching compared with original brute force approach.

Reply:
I'm ok with this optimization if it does not break anything. By eye-balling, it seems correct to me.